File: momentra/momentra/Shell/MomentExperience.swift
import Foundation

/// Presentation-level Moment experience — derived from real list reads, not a backend state machine.
enum MomentExperienceKind: Equatable {
    case loading
    case error
    case firstMoment
    case active
    case betweenMoments
    case pausedOnly
}

struct MomentSummary: Equatable, Identifiable {
    let momentId: String
    let title: String
    let status: String
    let momentTypeCode: String?
    /// Present for BUSINESS bootstrap moments — Company→Moment scoping.
    let companyId: String?
    var id: String { momentId }

    init(
        momentId: String,
        title: String,
        status: String,
        momentTypeCode: String? = nil,
        companyId: String? = nil
    ) {
        self.momentId = momentId
        self.title = title
        self.status = status
        self.momentTypeCode = momentTypeCode
        self.companyId = companyId
    }

    var isActiveStatus: Bool {
        status.caseInsensitiveCompare("ACTIVE") == .orderedSame ||
            status.caseInsensitiveCompare("DRAFT") == .orderedSame
    }

    var isPausedStatus: Bool {
        status.caseInsensitiveCompare("PAUSED") == .orderedSame
    }

    var isHistoricalStatus: Bool {
        status.caseInsensitiveCompare("COMPLETED") == .orderedSame ||
            status.caseInsensitiveCompare("CANCELLED") == .orderedSame ||
            status.caseInsensitiveCompare("ARCHIVED") == .orderedSame ||
            status.caseInsensitiveCompare("DELETED") == .orderedSame
    }
}

func resolveMomentExperience(_ moments: [MomentSummary]) -> MomentExperienceKind {
    if moments.isEmpty { return .firstMoment }
    if moments.contains(where: \.isActiveStatus) { return .active }
    let hasPaused = moments.contains(where: \.isPausedStatus)
    let hasHistory = moments.contains { $0.isHistoricalStatus || !$0.isPausedStatus }
    if hasPaused && !hasHistory { return .pausedOnly }
    return .betweenMoments
}

func recentHistoryMoments(_ moments: [MomentSummary], limit: Int = 5) -> [MomentSummary] {
    Array(
        moments
            .filter {
                !$0.isActiveStatus &&
                    $0.status.caseInsensitiveCompare("ARCHIVED") != .orderedSame
            }
            .prefix(limit)
    )
}

func activeMomentCount(_ moments: [MomentSummary]) -> Int {
    moments.filter(\.isActiveStatus).count
}

File: apk/app/src/main/java/com/example/momentra/domain/MomentExperience.kt
package com.example.momentra.domain

/** Presentation-level Moment experience — derived from real list reads, not a backend SM. */
enum class MomentExperienceKind {
    LOADING,
    ERROR,
    FIRST_MOMENT,
    ACTIVE,
    BETWEEN_MOMENTS,
    PAUSED_ONLY,
}

data class MomentSummary(
    val momentId: String,
    val title: String,
    val status: String,
    val momentTypeCode: String? = null,
    /** Present for BUSINESS bootstrap moments — used for Company→Moment scoping. */
    val companyId: String? = null,
    /** Present for GROUP bootstrap moments — active participant count. */
    val participantCount: Int = 0,
)

fun MomentSummary.isActiveStatus(): Boolean =
    status.equals("ACTIVE", ignoreCase = true) || status.equals("DRAFT", ignoreCase = true)

fun MomentSummary.isCompletedStatus(): Boolean =
    status.equals("COMPLETED", ignoreCase = true)

fun MomentSummary.isPausedStatus(): Boolean =
    status.equals("PAUSED", ignoreCase = true)

fun MomentSummary.isHistoricalStatus(): Boolean =
    status.equals("COMPLETED", ignoreCase = true) ||
        status.equals("CANCELLED", ignoreCase = true) ||
        status.equals("ARCHIVED", ignoreCase = true) ||
        status.equals("DELETED", ignoreCase = true)

/** Moments that can drive live shell product (Pulse/Finance), including settled COMPLETED. */
fun MomentSummary.isShellProductStatus(): Boolean =
    isActiveStatus() || isCompletedStatus()

fun resolveMomentExperience(moments: List<MomentSummary>): MomentExperienceKind {
    if (moments.isEmpty()) return MomentExperienceKind.FIRST_MOMENT
    val hasProduct = moments.any { it.isShellProductStatus() }
    if (hasProduct) return MomentExperienceKind.ACTIVE
    val hasPaused = moments.any { it.isPausedStatus() }
    val hasHistory = moments.any { it.isHistoricalStatus() || !it.isPausedStatus() }
    return when {
        hasPaused && !hasHistory -> MomentExperienceKind.PAUSED_ONLY
        else -> MomentExperienceKind.BETWEEN_MOMENTS
    }
}

fun recentHistoryMoments(moments: List<MomentSummary>, limit: Int = 5): List<MomentSummary> =
    moments
        .filter {
            !it.isActiveStatus() &&
                !it.status.equals("ARCHIVED", ignoreCase = true)
        }
        .take(limit)

fun activeMomentCount(moments: List<MomentSummary>): Int =
    moments.count { it.isActiveStatus() }
